### PR TITLE
Feature/animation loop optimization

### PR DESCRIPTION
**Pull request by bjonnh**
Allows me to get from 30fps to 50fps. And significantly reduce lag and
CPU usaqe on notebooks containing >10 molecules.

You can give it a try to see if it also improves for you.

### CODE REVIEW

**Comment by patrickfuller** [comment]
Makes sense to me - I'll look at it in more detail tomorrow.

If you're interested in performance, there is a significantly larger upgrade that helps. My [merged-geometry](https://github.com/patrickfuller/imolecule/tree/merged-geometry) branch is a huge performance boost on larger molecules ([ref](http://learningthreejs.com/blog/2011/10/05/performance-merging-geometry/)). I ended up not implementing it because I wasn't concerned with performance at the time. Might be worth dusting off.

File: js/imolecule.js [hunk 1]
```diff
@@ -334,7 +337,11 @@ var imolecule = {
             return self.animate();
         });
         this.controls.update();
-        this.renderer.render(this.scene, this.camera);
+    },
+
+    render: function () {
+        var self = this;
```
Comment (patrickfuller) [inline js/imolecule.js]: Why do you use `self` here?

File: js/imolecule.js [hunk 2]
```diff
@@ -307,6 +307,8 @@ var imolecule = {
 
     // Sets camera type (orthogonal, perspective)
     setCameraType: function (type) {
+        var self = this;
+
```
Comment (patrickfuller) [inline js/imolecule.js]: Watch your whitespace

**Comment by patrickfuller** [comment]
One minor bug: the current code will mess up on window resize:

screen shot 2015-03-27 at 10 15 29 am

You need to add `self.render()` into `$(window).resize` (line 46) to work.

**Comment by bjonnh** [comment]
Corrected all of that. Thanks.
I'm thinking about the merge-geometry. The patch cannot be applied as is as many things changed since. I was thinking about an optional geometry merge as some people may want to move the atoms (being able to see the bend, torsion related to an IR spectrum or whatever you can imagine). I'll try to work on that as soon as this one gets done.

**Comment by patrickfuller** [comment]
Looks good to me!

Thinking back to it, I had the same thoughts on merging geometry. You sacrifice functionality for performance, and you additionally increase the initial load time. It's worth keeping in mind if you're performance limited, but probably not needed in standard usage.

**Pull request merged**